cmd/internal/server/handlers/schema_aware_serializer.go

```diff
@@ -0,0 +1,229 @@
+package handlers
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/planetscale/fivetran-source/lib"
+
+	"github.com/pkg/errors"
+
+	fivetransdk "github.com/planetscale/fivetran-proto/go"
+	"vitess.io/vitess/go/sqltypes"
+)
+
+type Serializer interface {
+	Info(string)
+	Log(fivetransdk.LogLevel, string) error
+	Record(*sqltypes.Result, *fivetransdk.SchemaSelection, *fivetransdk.TableSelection) error
+	State(lib.SyncState) error
+	Release()
+}
+
+type LogSender interface {
+	Send(*fivetransdk.UpdateResponse) error
+}
+
+type schemaAwareSerializer struct {
+	prefix                 string
+	sender                 LogSender
+	recordResponseKey      string
+	recordResponse         *fivetransdk.UpdateResponse
+	serializeTinyIntAsBool bool
+	serializers            map[string]*recordSerializer
+	schemaList             *fivetransdk.SchemaList
+}
+
+type recordSerializer interface {
+	Serialize(result *sqltypes.Result) ([]map[string]*fivetransdk.ValueType, error)
+}
+
+type schemaAwareRecordSerializer struct {
+	columnSelection map[string]bool
+	columnWriters   map[string]func(value sqltypes.Value) (*fivetransdk.ValueType, error)
+}
+
+func (s *schemaAwareRecordSerializer) Serialize(result *sqltypes.Result) ([]map[string]*fivetransdk.ValueType, error) {
+	data := make([]map[string]*fivetransdk.ValueType, 0, len(result.Rows))
+	columns := make([]string, 0, len(result.Fields))
+	for _, field := range result.Fields {
+		columns = append(columns, field.Name)
+	}
+
+	for _, row := range result.Rows {
+		record := make(map[string]*fivetransdk.ValueType)
+		for idx, val := range row {
+			if idx < len(columns) {
```

Comment:
Invert this check and break in the if statement to make it more obvious that this terminates scanning columns.